File: linux/gamelift/gamelift-upload/base/share.lua
local jsEvents = require 'jsEvents'

local share = {}

function share.connectClient(mediaUrl, callback)
    jsEvents.listen('CASTLE_CONNECT_MULTIPLAYER_CLIENT_RESPONSE', function(params)
        callback(params.address)
    end)
    jsEvents.send('CASTLE_CONNECT_MULTIPLAYER_CLIENT_REQUEST', {
        mediaUrl = mediaUrl,
    })
end

return share

File: web-source/common/share.js
import * as Actions from '~/common/actions';
import * as CEF from '~/common/cef';

class Share {
  addEventListeners = () => {
    window.addEventListener(
      'CASTLE_CONNECT_MULTIPLAYER_CLIENT_REQUEST',
      this._connectMultiplayerClientAsync
    );
  };

  removeEventListeners = () => {
    window.removeEventListener(
      'CASTLE_CONNECT_MULTIPLAYER_CLIENT_REQUEST',
      this._connectMultiplayerClientAsync
    );
  };

  _connectMultiplayerClientAsync = async e => {
    let mediaUrl = e.params.mediaUrl;
    let response = await Actions.multiplayerJoinAsync(mediaUrl);

    CEF.sendLuaEvent('CASTLE_CONNECT_MULTIPLAYER_CLIENT_RESPONSE', {
      address: response.address,
    });
  };
}

export default new Share();
